File: AndroidNative/games/src/main/java/com/chathil/games/data/remote/response/GameResponse.kt
package com.chathil.games.data.remote.response

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class GameResponse(
    @SerialName("id")
    val id: Int,
    @SerialName("slug")
    val slug: String? = null,
    @SerialName("name")
    val name: String,
    @SerialName("released")
    val released: String?,
    @SerialName("tba")
    val tba: Boolean,
    @SerialName("background_image")
    val backgroundImage: String? = null,
    @SerialName("rating")
    val rating: Float,
    @SerialName("rating_top")
    val ratingTop: Float,
    @SerialName("ratings")
    val ratings: List<RatingResponse>? = emptyList(),
    @SerialName("ratings_count")
    val ratingsCount: Int,
    @SerialName("reviews_text_count")
    val reviewsTextCount: Int,
    @SerialName("added")
    val added: Int,
    @SerialName("added_by_status")
    val addedByStatus: AddedByStatus?,
    @SerialName("metacritic")
    val metacritic: Int? = 0,
    @SerialName("playtime")
    val playtime: Int,
    @SerialName("suggestions_count")
    val suggestionsCount: Int,
    @SerialName("reviews_count")
    val reviewsCount: Int,
    @SerialName("saturated_color")
    val saturatedColor: String,
    @SerialName("dominant_color")
    val dominantColor: String,
    @SerialName("platforms")
    val platforms: List<GamePlatformResponse>? = emptyList(),
    @SerialName("parent_platforms")
    val parentPlatforms: List<GameParentPlatformResponse>? = emptyList(),
    @SerialName("genres")
    val genres: List<GenreResponse>? = emptyList(),
    @SerialName("stores")
    val stores: List<GameStoreResponse>? = emptyList(),
    @SerialName("clip")
    val clip: ClipResponse?,
    @SerialName("tags")
    val tags: List<TagResponse>? = emptyList(),
    @SerialName("short_screenshots")
    val shortScreenshots: List<ShortScreenshotResponse>? = emptyList()
)

@Serializable
data class RatingResponse(
    @SerialName("id")
    val id: Int,
    @SerialName("title")
    val title: String,
    @SerialName("count")
    val count: Int,
    @SerialName("percent")
    val percent: Float
)

@Serializable
data class PlatformResponse(
    @SerialName("id")
    val id: Int,
    @SerialName("name")
    val name: String,
    @SerialName("slug")
    val slug: String? = null,
    @SerialName("image")
    val image: String? = null,
    @SerialName("year_end")
    val yearEnd: Int? = null,
    @SerialName("year_start")
    val yearStart: Int? = null,
    @SerialName("games_count")
    val gamesCount: Int? = 0,
    @SerialName("image_background")
    val imageBackground: String? = null,
)

@Serializable
data class RequirementResponse(
    @SerialName("minimum")
    val minimum: String? = null,
    @SerialName("recommended")
    val recommended: String? = null
)

@Serializable
data class GamePlatformResponse(
    @SerialName("platform")
    val platform: PlatformResponse? = null,
    @SerialName("released_at")
    val releasedAt: String? = null,
    @SerialName("requirements_en")
    val requirementsEn: RequirementResponse? = null
)

@Serializable
data class ParentPlatformResponse(
    @SerialName("id")
    val id: Int,
    @SerialName("name")
    val name: String,
    @SerialName("slug")
    val slug: String? = null
)

@Serializable
data class GameParentPlatformResponse(
    @SerialName("platform")
    val platform: ParentPlatformResponse
)

@Serializable
data class GenreResponse(
    @SerialName("id")
    val id: Int,
    @SerialName("name")
    val name: String,
    @SerialName("slug")
    val slug: String? = null,
    @SerialName("games_count")
    val gamesCount: Int? = 0,
    @SerialName("image_background")
    val imageBackground: String? = null
)

@Serializable
data class StoreResponse(
    @SerialName("id")
    val id: Int,
    @SerialName("name")
    val name: String,
    @SerialName("slug")
    val slug: String? = null,
    @SerialName("domain")
    val domain: String? = null,
    @SerialName("games_count")
    val gamesCount: Int? = 0,
    @SerialName("image_background")
    val imageBackground: String? = null
)

@Serializable
data class GameStoreResponse(
    @SerialName("store")
    val store: StoreResponse,
    @SerialName("url_en")
    val urlEn: String? = null
)

@Serializable
data class ClipResponse(
    @SerialName("clip")
    val clip: String? = null
)

@Serializable
data class TagResponse(
    @SerialName("id")
    val id: Int,
    @SerialName("name")
    val name: String,
    @SerialName("slug")
    val slug: String? = null,
    @SerialName("language")
    val language: String? = null,
    @SerialName("games_count")
    val gamesCount: Int? = 0,
    @SerialName("image_background")
    val imageBackground: String? = null
)

@Serializable
data class ShortScreenshotResponse(
    @SerialName("id")
    val id: Int,
    @SerialName("image")
    val image: String? = null
)

@Serializable
data class AddedByStatus(
    val yet: Int = 0,
    val owned: Int = 0,
    val beaten: Int = 0,
    val toplay: Int = 0,
    val dropped: Int = 0,
    val playing: Int = 0
)

File: iOSNative/Games/Sources/Games/Data/Remote/Response/GameResponse.swift
//
//  File.swift
//  
//
//  Created by Abdul Chathil on 30/03/21.
//

import Foundation

public struct GameResponse: Codable, Identifiable {
  public let id: Int32
  public let slug: String?
  public let name: String
  public let released: String?
  public let tba: Bool
  public let backgroundImage: String?
  public let rating: Double?
  public let ratingTop: Double?
  public let ratings: [RatingResponse]
  public let ratingsCount: Int
  public let reviewsTextCount: Int
  public let added: Int
  public let metacritic: Int16?
  public let playtime: Int
  public let suggestionsCount: Int
  public let reviewsCount: Int
  public let saturatedColor: String
  public let dominantColor: String
  public let platforms: [GamePlatformResponse]
  public let parentPlatforms: [GameParentPlatformResponse]
  public let genres: [GenreResponse]
  public let stores: [GameStoreResponse]
  public let clip: ClipResponse?
  public let tags: [TagResponse]
  public let shortScreenshots: [ShortScreenshotResponse]
  
  private enum CodingKeys: String, CodingKey {
    case id = "id"
    case name = "name"
    case slug = "slug"
    case released = "released"
    case tba = "tba"
    case backgroundImage = "background_image"
    case rating = "rating"
    case ratingTop = "rating_top"
    case ratings = "ratings"
    case ratingsCount = "ratings_count"
    case reviewsTextCount = "reviews_text_count"
    case added = "added"
    case metacritic = "metacritic"
    case playtime = "playtime"
    case suggestionsCount = "suggestions_count"
    case reviewsCount = "reviews_count"
    case saturatedColor = "saturated_color"
    case dominantColor = "dominant_color"
    case platforms = "platforms"
    case parentPlatforms = "parent_platforms"
    case genres = "genres"
    case stores = "stores"
    case clip = "clip"
    case tags = "tags"
    case shortScreenshots = "short_screenshots"
  }
  
}

public struct RatingResponse: Codable {
  public let id: Int
  public let title: String
  public let count: Int
  public let percent: Float
}

public struct GenreResponse: Codable {
  public let id: Int
  public let name: String
  public let slug: String?
  public let gamesCount: Int?
  public let imageBackground: String?
  private enum CodingKeys: String, CodingKey {
    case id = "id"
    case name = "name"
    case slug = "slug"
    case gamesCount = "games_count"
    case imageBackground = "image_background"
  }
}

public struct PlatformResponse: Codable {
  public let id: Int
  public let name: String
  public let slug: String?
  public let image: String?
  public let yearEnd: Int?
  public let yearStart: Int?
  public let gamesCount: Int?
  public let imageBackground: String
  private enum CodingKeys: String, CodingKey {
    case id = "id"
    case name = "name"
    case slug = "slug"
    case image = "image"
    case yearEnd = "year_end"
    case yearStart = "year_start"
    case gamesCount = "games_count"
    case imageBackground = "image_background"
  }
}

public struct RequirementResponse: Codable {
  public let minimum: String?
  public let recommended: String
}

public struct GamePlatformResponse: Codable {
  public let platform: PlatformResponse?
  public let releasedAt: String?
  public let requirementsEn: RequirementResponse?
  private enum CodingKeys: String, CodingKey {
    case platform = "platform"
    case releasedAt = "released_at"
    case requirementsEn = "requirements_en"
  }
}

public struct ParentPlatformResponse: Codable {
  public let id: Int
  public let name: String
  public let slug: String?
}

public struct GameParentPlatformResponse: Codable {
  public let platform: ParentPlatformResponse
}

public struct StoreResponse: Codable {
  public let id: Int
  public let name: String
  public let slug: String?
  public let domain: String?
  public let gamesCount: Int
  public let imageBackground: String?
  private enum CodingKeys: String, CodingKey {
    case id = "id"
    case name = "name"
    case slug = "slug"
    case domain = "domain"
    case gamesCount = "games_count"
    case imageBackground = "image_background"
  }
}

public struct GameStoreResponse: Codable {
  public let store: StoreResponse
  public let urlEn: String?
  private enum CodingKeys: String, CodingKey {
    case store = "id"
    case urlEn = "name"
  }
}

public struct ClipResponse: Codable {
  public let clip: String?
}

public struct TagResponse: Codable {
  public let id: Int
  public let name: String
  public let slug: String?
  public let language: String?
  public let gamesCount: Int?
  public let imageBackground: String?
  private enum CodingKeys: String, CodingKey {
    case id = "id"
    case name = "name"
    case slug = "slug"
    case language = "language"
    case gamesCount = "games_count"
    case imageBackground = "image_background"
  }
}

public struct ShortScreenshotResponse:Codable {
  public let id: Int
  public let image: String?
}
